### ISSUE TITLE
[MEND] Fast Model Editing at Scale

### ISSUE THREAD

**Issue by bigshanedogg** [bigshanedogg/survey#40]
## Problem statement

1. model edit 방법이 현재 시점에서 실용적이려면 large-scale 모델에서도 feasible time 내에 적용이 가능해야 한다.
  - generality와 locality 사이의 균형적인 상승을 가져오는 model edit 방법을 제안한다. (KnowledgeEditor에서 제안한 문제점과 동일하다.)
  - KE에서 연산 비용 상으로 조금 더 개선된 model edit 방법으로 11B (T5-XXL)까지의 scalability 확인한다.

## Baseline
- FT
- FT+KL: KL divergence loss로 parameter 업데이트를 제한하는 FT
- KE (KnowledgeEditor)

## Data details
- WikiText
- FEVER
- zsRE (zero-shot Relation Extraction)

## Approach
- MEND (Model Editor Networks with Gradient Decomposition): fine-tuning gradients를 targeted parameter update로 변환할 수 있는 방법론
  - ![Screen Shot 2024-09-11 at 10 37 20 AM](https://github.com/user-attachments/assets/7758e002-fbf5-41b3-b3a0-ce912cc4c944)
  - MEND는 θ'로의 직접적인 변형이 아닌 decomposition을 사용한다. 
    - 아래 수식과 같이 θ'를 분해할 수 있으므로 backprop 시의 직전 레이어 (l+1)의 그라디언트 δ와 현재 레이어의 입력인 u를 입력으로 받아서 δ'와 u'를 예측하도록 하고, 이를 이용해 θ'를 계산하는 식이다. 
    - ![Screen Shot 2024-09-11 at 1 18 59 PM](https://github.com/user-attachments/assets/e8e78cc8-c6c0-44c5-bb41-47ebdb7efb4e)
    - 이는 레이어의 matrix의 크기에 quadratic하게 증가하는 기존 방식 대비, linear한 비용만 증가시킨다.
    - KE와의 차이점: KE는 본래 모델의 그라디언트 θ을 θ'로 변환시킬 계수와 편차를 구하는 방식을 사용한다. 이 때, `<x,y,a>`를 LSTM으로 인코딩한 h에 대해 계수와 편차를 구하기 위한 4개의 벡터와 1개의 스칼라를 예측하는 5개의 FFNN을 사용하는 식이므로, MEND에 비해 computational cost가 더 높다.
  - MEND의 학습은 큰 골자는 KE와 유사하다. parameter update를 위한 loss와 locality를 위한 loss로 구성되어 있다.
    - ![Screen Shot 2024-09-11 at 1 20 21 PM](https://github.com/user-attachments/assets/81164d6b-a1bc-465b-ba24-7f7213d1d488)
- MEND는 아래와 같이 기존 베이스라인 대비 scale-up과 batched edit이 가능하다는 점에서 차별점이 있다.
  - ![Screen Shot 2024-09-11 at 1 22 44 PM](https://github.com/user-attachments/assets/73b4c661-f0d8-49b3-93e3-3d5003fc653c)

### 정량 결과
- ![Screen Shot 2024-09-11 at 1 30 32 PM](https://github.com/user-attachments/assets/3a770832-d865-432e-900e-c74ac14ce644)
  - small-scale LM에 대해선 FT를 제외한 기존 baselines과의 현저한 차이는 보이지 않는다. 대체로, 높은 generality (높은 edit success)와 높은 reliability (낮은 acc drawdown)를 보여준다.
- ![Screen Shot 2024-09-11 at 1 33 06 PM](https://github.com/user-attachments/assets/109abc7c-06bc-4a3a-9094-66a0d404026c)
  - large-scale에 대해서 일부 방법론 (ENN)은 과도한 연산 비용 때문에 실험이 어려우며, KE의 경우에는 매우 낮은 edit success를 보인다. MEND는 그 외 FT류의 베이스라인과 비교했을 때 generality와 reliability 측면에서 각각 모두 더 높은 성능을 보인다.

### 정성 샘플
- ![Screen Shot 2024-09-11 at 1 24 04 PM](https://github.com/user-attachments/assets/a6d4506e-ff98-47b8-83d7-7851e56ea31c)
  - 아래 부분의 failure example도 있고 어느 정도 체리픽이 있을 것이라 생각하지만, re-training 없이 작은 규모의 hyper-network만으로 이 정도의 knowledge update가 이루어진다는 점에서 다른 어플리케이션에도 실험해볼만한 것 같다.
  - 기존의 논문들이 QA나 이진분류(FC)에 대한 실험이 주였다면, WikiText와 같이 generated task에 대해서도 실험했다는 점에서 더 활용가치가 높지 않을까.
